File: web_port/core/SmoothInterpolator.ts
/**
 * SmoothInterpolator.ts
 * 
 * Portage TypeScript du système d'interpolation fluide de SimGolf.
 * 
 * Source ASM : golf.exe 0x4969e0, 0x496c30, 0x496d60, 0x496e30
 * 
 * Ce système animait la caméra, le zoom, les transitions UI et
 * potentiellement la trajectoire de balle.
 * 
 * Principe : une valeur numérique bouge entre min et max avec
 * accélération/décélération, notifiée via callback.
 * 
 * Utilisation :
 *   const anim = new SmoothInterpolator(0, 100, 10);
 *   anim.onUpdate = (pos) => camera.setZoom(pos);
 *   anim.target = 80;     // va bouger de 0 à 80 avec step 10
 *   anim.update(deltaMs); // appelé chaque frame
 * 
 * @package simgolf.core
 */

/** Callback de notification de changement de position */
export type PositionCallback = (position: number) => void;

/**
 * SmoothInterpolator
 * 
 * Animation fluide entre deux bornes (min → max).
 * Supporte l'accélération directionnelle et les seuils
 * de déclenchement.
 * 
 * Mode d'emploi :
 * 1. Créer l'interpolateur avec min, max
 * 2. Définir onUpdate (callback appelé à chaque changement)
 * 3. Définir les seuils (trigger points pour changement de direction)
 * 4. Appeler update(deltaMs) chaque frame
 */
export class SmoothInterpolator {
    // ============================================================
    // Membres (correspondent aux offsets de la structure C)
    // ============================================================

    /** Identifiant de la cible (ex: "camera-zoom" ou "golfer-3") */
    targetId: string = '';

    /** Borne inférieure */
    minPosition: number = 0;

    /** Borne supérieure */
    maxPosition: number = 100;

    /** Position courante (interpolée) */
    currentPosition: number = 0;

    /** Durée totale de l'animation en ms */
    duration: number = 1000;

    /** Timestamp de la dernière mise à jour (performance.now()) */
    private lastTimestamp: number = 0;

    /** Flag : inverser la direction de notification */
    directionInverted: boolean = false;

    /** Pas de déplacement par tick */
    stepSize: number = 1;

    /** 
     * Accélération directionnelle
     *  1 → se déplace vers max
     * -1 → se déplace vers min
     *  0 → arrêté
     */
    acceleration: number = 0;

    /** Valeur de déclenchement haute */
    upperTrigger: number = 80;

    /** Valeur de déclenchement basse */
    lowerTrigger: number = 20;

    /** Position cible haute */
    upperTarget: number = 100;

    /** Position cible basse */
    lowerTarget: number = 0;

    // ============================================================
    // Callbacks
    // ============================================================

    /** 
     * Callback principal — appelé quand la position change.
     * Reçoit la nouvelle position interpolée.
     */
    onUpdate: PositionCallback | null = null;

    /**
     * Callback secondaire — appelé après onUpdate.
     * Reçoit également la nouvelle position.
     */
    onUpdateSecondary: PositionCallback | null = null;

    /**
     * Callback de fin d'animation — appelé quand 
     * l'interpolation est terminée (position stable).
     */
    onComplete: (() => void) | null = null;

    /** Flag indiquant si l'animation est en cours */
    private _isAnimating: boolean = false;

    // ============================================================
    // Constructeur
    // ============================================================

    constructor(
        minPosition: number = 0,
        maxPosition: number = 100,
        stepSize: number = 1,
        duration: number = 1000
    ) {
        this.minPosition = minPosition;
        this.maxPosition = maxPosition;
        this.stepSize = stepSize;
        this.duration = duration;
        this.currentPosition = minPosition;
    }

    // ============================================================
    // Méthodes principales
    // ============================================================

    /**
     * Met à jour l'interpolation.
     * 
     * Logique originale (ASM 0x4969e0) :
     * 1. Calcule le delta depuis la dernière frame
     * 2. Si le delta dépasse la moitié de la durée → idle
     * 3. Sinon, applique le step dans la direction courante
     * 4. Vérifie les seuils pour décider du changement de direction
     * 5. Notifie via callback
     * 
     * @param deltaMs Temps écoulé depuis la dernière frame (ms)
     */
    update(deltaMs: number): void {
        // --- Phase 1 : Vérification du delta ---
        if (this.lastTimestamp > 0) {
            const halfDuration = this.duration * 2;

            if (deltaMs >= halfDuration) {
                // Trop de temps écoulé → arrêt et notification
                this.acceleration = 0;
                this.notifyCallbacks();
                return;
            }
        } else {
            // Première frame : initialise le timestamp
            this.lastTimestamp = performance.now();
        }

        // --- Phase 2 : Calcul de la nouvelle position ---
        // Détermine si on est dans une phase de dépassement 
        // (hasCurrent > hasEnd d'après l'ASM)
        const hasCurrent = this.acceleration;  // proxy pour timing_source
        const hasEnd = this.currentPosition >= this.upperTarget ? 1 : 0;

        if (this.currentPosition > this.upperTrigger) {
            // Au-dessus du seuil haut
            if (this.acceleration <= 0) {
                // Pas encore en mouvement vers le bas → déclenche
                this.acceleration = 1;      // vers max
                this.lastTimestamp = 0;
                this.applyStep(deltaMs);
                this.notifyCallbacks();
                return;
            }
        } else if (this.currentPosition < this.lowerTrigger) {
            // En dessous du seuil bas
            if (this.acceleration >= 0) {
                this.acceleration = -1;     // vers min
                this.lastTimestamp = 0;
                this.applyStep(deltaMs);
                this.notifyCallbacks();
                return;
            }
        }

        // Aucun changement de direction → applique le step normal
        this.applyStep(deltaMs);

        // --- Phase 3 : Notification ---
        this.notifyCallbacks();
    }

    /**
     * Applique le pas de déplacement dans la direction courante,
     * avec clamping aux bornes.
     */
    private applyStep(deltaMs: number): void {
        let newPosition: number;

        if (this.acceleration <= 0) {
            // Se déplace vers max
            newPosition = this.currentPosition + this.stepSize;
            newPosition = Math.min(newPosition, this.maxPosition);
        } else {
            // Se déplace vers min
            newPosition = this.currentPosition - this.stepSize;
            newPosition = Math.max(newPosition, this.minPosition);
        }

        this.currentPosition = newPosition;
    }

    /**
     * Notifie les callbacks du changement de position.
     * 
     * Logique originale (ASM 0x496c06-0x496c25) :
     * - Appelle vtable+0xe8 (onUpdate principal)
     * - Appelle le callback personnalisé si présent
     * - Appelle vtable+0x120 (onComplete)
     * - Met à jour le timestamp
     */
    private notifyCallbacks(): void {
        // Position à notifier (dépend du mode inverted)
        const notifyPosition = this.directionInverted
            ? (this.minPosition - this.currentPosition) + this.maxPosition
            : this.currentPosition;

        // Callback principal
        if (this.onUpdate) {
            this.onUpdate(notifyPosition);
        }

        // Callback secondaire
        if (this.onUpdateSecondary) {
            this.onUpdateSecondary(notifyPosition);
        }

        // Vérifie si la position a changé
        const positionChanged = this.currentPosition !== 
            (this._lastNotifiedPosition ?? this.currentPosition);
        
        if (!positionChanged && this.acceleration === 0) {
            // Animation terminée
            if (this.onComplete) {
                this.onComplete();
            }
            this._isAnimating = false;
        }

        // Met à jour le timestamp
        this.lastTimestamp = performance.now();
        this._lastNotifiedPosition = this.currentPosition;
    }

    private _lastNotifiedPosition: number | null = null;

    // ============================================================
    // Méthodes de contrôle
    // ============================================================

    /**
     * Définit une cible et démarre l'animation.
     * 
     * Équivalent ASM : FUN_0x496c30
     * 
     * @param current Valeur actuelle (point de départ)
     * @param target Valeur cible
     */
    setTarget(current: number, target: number): void {
        if (target > this.upperTarget) {
            // Cible au-dessus → accélération vers max
            this.acceleration = 1;
            this.lastTimestamp = 0;
        } else if (target < this.lowerTarget) {
            // Cible en dessous → accélération vers min
            this.acceleration = -1;
            this.lastTimestamp = 0;
        } else {
            // Cible dans la zone → accélération selon direction
            this.acceleration = current < this.upperTrigger ? -1 : 1;
            this.lastTimestamp = 0;
        }

        this._isAnimating = true;
        this.applyStep(0);  // Applique immédiatement un step
        this.notifyCallbacks();
    }

    /**
     * Déplace directement à une position spécifique.
     * 
     * Équivalent ASM : FUN_0x496e30
     * Modes : 
     *   'max'     → position = maxPosition
     *   'min'     → position = minPosition  
     *   'step-up' → currentPosition += stepSize (clamp max)
     *   'step-dn' → currentPosition -= stepSize (clamp min)
     * 
     * @param mode Mode de déplacement
     */
    setPositionDirect(mode: 'max' | 'min' | 'step-up' | 'step-dn'): void {
        switch (mode) {
            case 'max':
                this.currentPosition = this.maxPosition;
                break;
            case 'min':
                this.currentPosition = this.minPosition;
                break;
            case 'step-up': {
                const newPos = this.currentPosition + this.stepSize;
                this.currentPosition = Math.min(newPos, this.maxPosition);
                break;
            }
            case 'step-dn': {
                const newPos = this.currentPosition - this.stepSize;
                this.currentPosition = Math.max(newPos, this.minPosition);
                break;
            }
        }

        this._isAnimating = false;
        this.acceleration = 0;
        this.notifyCallbacks();
    }

    /**
     * Réinitialise l'interpolateur.
     */
    reset(): void {
        this.currentPosition = this.minPosition;
        this.acceleration = 0;
        this.lastTimestamp = 0;
        this._isAnimating = false;
        this._lastNotifiedPosition = null;
    }

    /**
     * Vrai si une animation est en cours.
     */
    get isAnimating(): boolean {
        return this._isAnimating;
    }

    /**
     * Retourne la progression normalisée (0.0 - 1.0).
     */
    get progress(): number {
        const range = this.maxPosition - this.minPosition;
        if (range === 0) return 1;
        return (this.currentPosition - this.minPosition) / range;
    }

    /**
     * Chaîne de debug.
     */
    toString(): string {
        return `SmoothInterpolator(${this.targetId}) ` +
            `[${this.currentPosition}/${this.maxPosition}] ` +
            `accel=${this.acceleration} step=${this.stepSize} ` +
            `triggers=[${this.lowerTrigger},${this.upperTrigger}]`;
    }
}

// ============================================================
// Tests unitaires (node.js)
// ============================================================

if (require && require.main === module) {
    // Test basique : animation de zoom caméra
    const zoom = new SmoothInterpolator(0, 100, 5, 500);
    zoom.targetId = 'camera-zoom';
    zoom.upperTrigger = 70;
    zoom.lowerTrigger = 30;

    let lastPos = 0;
    zoom.onUpdate = (pos) => {
        lastPos = pos;
        console.log(`  Zoom: ${pos} (accel=${zoom.acceleration})`);
    };

    console.log('=== Test : Animation zoom caméra ===');
    console.log(`État initial: ${zoom}`);

    // Simule 20 frames d'animation
    for (let frame = 0; frame < 20; frame++) {
        zoom.update(50);  // 50ms par frame
    }

    console.log(`\nÉtat final: ${zoom}`);

    // Test : déplacement direct
    console.log('\n=== Test : Déplacement direct ===');
    zoom.reset();
    zoom.setPositionDirect('max');
    console.log(`Position directe max: ${zoom}`);

    // Test : mode inversé
    console.log('\n=== Test : Mode inversé ===');
    const inverted = new SmoothInterpolator(0, 360, 10, 1000);
    inverted.targetId = 'camera-rotation';
    inverted.directionInverted = true;
    inverted.onUpdate = (pos) => {
        console.log(`  Rotation (inverted): ${pos.toFixed(1)}°`);
    };
    inverted.setTarget(45, 270);
    for (let frame = 0; frame < 15; frame++) {
        inverted.update(50);
    }
}

File: ref/decompiled/smooth_interpolator.c
/**
 * smooth_interpolator.c
 * 
 * Source : golf.exe dépaqueté (DEViANCE) — FUN_0x4969e0
 * 
 * Système d'interpolation fluide pour caméra, position, zoom.
 * Déplacé de l'ASM à ~0x4969e0 (fonction smooth update avec timing).
 * 
 * Analyse Hermes-RE, Mai 2026
 */

#include <stdint.h>
#include <stdbool.h>

// ============================================================
// Structures de données
// ============================================================

/**
 * SmoothInterpolator — Animateur fluide entre deux bornes
 * 
 * Fait bouger une valeur entre `min_pos` et `max_pos` avec
 * accélération/décélération.
 * 
 * Utilisé pour : zoom caméra, rotation, panoramique, 
 * transitions UI, animations balle.
 */
typedef struct SmoothInterpolator {
    // VTable (C++ virtuel) — pointeur vers la table des méthodes
    void** vtable;

    // === MEMBRES À OFFSET FIXE (depuis analyse golf.exe) ===
    
    // +0x578: Identifiant/cible de l'objet à mettre à jour
    //         (passé comme paramètre aux callbacks vtable+0xe4/e8)
    int target_id;

    // +0x580: Borne inférieure (minimum position)
    int min_pos;

    // +0x584: Borne supérieure (maximum position) 
    int max_pos;

    // +0x588: Direction inversée ?
    //         0 = forward (min→max), non-zero = backward (max→min)
    bool direction_inverted;

    // +0x58c: Position courante (interpolée)
    int current_pos;

    // +0x590: Durée totale de l'animation (en frames/ticks)
    int duration;

    // +0x594: (unknown — padding/autre usage)
    int unknown_594;

    // +0x598: Timestamp de la dernière mise à jour (timeGetTime)
    uint32_t last_timestamp;

    // +0x59c: État/statut de l'interpolation
    //         -1 = idle, 0 = en cours vers max, 1 = en cours vers min
    int status;

    // +0x5a0: Pas de déplacement par tick
    int step_size;

    // +0x5a4: Accélération directionnelle
    //         1 = se déplace vers max, -1 = vers min, 0 = arrêté
    int acceleration;

    // +0x5a8: (unknown)
    int unknown_5a8;

    // +0x5ac: Seuil haut pour déclencher la direction
    int upper_trigger;

    // +0x5b0: Seuil bas pour déclencher la direction  
    int lower_trigger;

    // +0x5b4: Cible haute
    int upper_target;

    // +0x5b8: Cible basse
    int lower_target;

    // +0x5bc: Limite haute ajustée
    int adjusted_upper;

    // +0x130: Objet cible pour les callbacks
    //         Contient vtable utilisée pour vtable+0xe4/e8
    void* target_object;

    // +0x278: Source de timing (objet avec vtable+0xd8=current, +0xdc=end)
    void* timing_source;

    // +0x13ac: Callback personnalisé (optionnel)
    //          Fonction appelée après mise à jour position
    void (*custom_callback)(int target_id, int new_position);

    // +0x13b0: Callback personnalisé (optionnel, v2)
    void (*custom_callback2)(int target_id, int new_position);

} SmoothInterpolator;

// ============================================================
// Méthodes (convention __thiscall : this dans ECX)
// ============================================================

/**
 * SmoothInterpolator_update
 * 
 * Met à jour la position avec interpolation temporelle.
 * 
 * Logique :
 * 1. Vérifie le delta depuis la dernière frame
 * 2. Si pas assez de temps passé, applique step vers target
 * 3. Clampe entre min_pos et max_pos
 * 4. Notifie l'objet cible via callback
 * 
 * @param this  L'interpolateur (this en __thiscall)
 * @param a     Premier paramètre (threshold/trigger)
 * @param b     Second paramètre (target)
 * 
 * Équivalent ASM : FUN_0x4969e0
 */
void SmoothInterpolator_update(SmoothInterpolator* this, 
                                int a, int b) {
    uint32_t now;
    int delta;
    int has_current, has_end;
    int temp_pos;

    // Récupère le timestamp actuel
    now = timeGetTime();  // [0x4ba21c]

    // Calcule le delta depuis la dernière mise à jour
    if (this->last_timestamp != 0) {
        delta = (int)(now - this->last_timestamp);
        int half_duration = this->duration * 2;  // shl 1

        // Si le delta a dépassé la moitié de la durée
        if (delta >= half_duration) {
            // Passe en mode idle et sort
            this->acceleration = 0;
            SmoothInterpolator_callInternal(this, &a, &b);
            return;
        }
    } else {
        // Première frame : initialise le timestamp
        this->last_timestamp = now;
    }

    // --- Phase 2 : Animation contrôlée ---
    // Récupère les valeurs de timing depuis l'objet source
    has_current = getTimingCurrent(this->timing_source);  // vtable+0xd8
    has_end     = getTimingEnd(this->timing_source);      // vtable+0xdc

    if (has_current > has_end) {
        // Cas LENGTH > ACCURACY (ou équivalent)
        if (a > this->upper_trigger) {
            // Au-dessus du seuil haut → accélération positive
            if (this->acceleration <= 0) {
                this->acceleration = -1;
                this->last_timestamp = 0;
                // Appelle le callback de déplacement
                this->vtable[0x6c >> 2](this, a, b);  // vtable+0x6c
                return;
            }
        } else if (a < this->lower_trigger) {
            // En dessous du seuil bas → accélération négative
            if (this->acceleration >= 0) {
                this->acceleration = 1;
                this->last_timestamp = 0;
                this->vtable[0x6c >> 2](this, a, b);
                return;
            }
        }

        // Aucun changement de direction → applique le step
        if (this->acceleration <= 0) {
            temp_pos = this->current_pos + this->step_size;
            if (temp_pos <= this->max_pos) {
                this->current_pos = temp_pos;
            } else {
                this->current_pos = this->max_pos;
            }
        } else {
            temp_pos = this->current_pos - this->step_size;
            if (temp_pos >= this->min_pos) {
                this->current_pos = temp_pos;
            } else {
                this->current_pos = this->min_pos;
            }
        }
    } else {
        // Cas ACCURACY >= LENGTH (inverse)
        if (a > this->upper_trigger) {
            if (this->acceleration <= 0) {
                this->acceleration = 1;
                this->last_timestamp = 0;
                this->vtable[0x6c >> 2](this, a, b);
                return;
            }
        } else if (a < this->lower_trigger) {
            if (this->acceleration >= 0) {
                this->acceleration = -1;
                this->last_timestamp = 0;
                this->vtable[0x6c >> 2](this, a, b);
                return;
            }
        }

        // Applique le step dans la direction appropriée
        if (this->acceleration <= 0) {
            temp_pos = this->current_pos + this->step_size;
            if (temp_pos <= this->max_pos) {
                this->current_pos = temp_pos;
            } else {
                this->current_pos = this->max_pos;
            }
        } else {
            temp_pos = this->current_pos - this->step_size;
            if (temp_pos >= this->min_pos) {
                this->current_pos = temp_pos;
            } else {
                this->current_pos = this->min_pos;
            }
        }
    }

    // --- Phase 3 : Finalisation et callback ---
    SmoothInterpolator_callInternal(this, &a, &b);
}

/**
 * SmoothInterpolator_callInternal
 * 
 * Notifie l'objet cible du changement de position.
 * Appelle vtable+0xe8 puis le callback optionnel.
 * 
 * @param this  L'interpolateur
 * @param a     Pointeur vers paramètre A (peut être modifié)
 * @param b     Pointeur vers paramètre B
 */
void SmoothInterpolator_callInternal(SmoothInterpolator* this,
                                      int* a, int* b) {
    // Met à jour le global "current interpolator" 
    // (utilisé par le système de rendu pour suivre l'animation)
    *((int*)0x83ab2c) = (int)this->target_object;

    // Vérifie si la position a changé
    if (*b != this->current_pos) {
        // Notifie l'objet cible
        void* target = this->target_object;
        void** target_vtable = *(void***)target;

        if (this->direction_inverted) {
            // Mode inversé : envoie (min - current + max)
            int inverted = (this->min_pos - this->current_pos) + this->max_pos;
            target_vtable[0xe8 >> 2](target, this->target_id, inverted);
        } else {
            // Mode normal : envoie la position courante
            target_vtable[0xe8 >> 2](target, this->target_id, this->current_pos);
        }

        // Callback personnalisé optionnel
        if (this->custom_callback2) {
            int callback_val;
            if (this->direction_inverted) {
                callback_val = (this->min_pos - this->current_pos) + this->max_pos;
            } else {
                callback_val = this->current_pos;
            }
            this->custom_callback2(this->target_id, callback_val);
        }
    }

    // Callback de fin d'animation (vtable+0x120)
    this->vtable[0x120 >> 2](this);

    // Met à jour le timestamp de la dernière frame
    this->last_timestamp = timeGetTime();
    this->status = -1;  // idle
}
